src/graph/on_disk_graph.cpp

```diff
@@ -225,26 +240,18 @@ Graph::Iterator OnDiskGraph::scanBwd(nodeID_t nodeID, GraphScanState& state) {
     return Graph::Iterator(&onDiskScanState);
 }
 
-std::vector<nodeID_t> OnDiskGraph::scanBwdRandom(nodeID_t, GraphScanState&) {
-    // auto& onDiskScanState = ku_dynamic_cast<OnDiskGraphScanStates&>(state);
-    // KU_ASSERT(nodeTableIDToBwdRelTables.contains(nodeID.tableID));
-    // auto& relTables = nodeTableIDToBwdRelTables.at(nodeID.tableID);
-    std::vector<nodeID_t> result;
-    // for (auto& [tableID, scanState] : onDiskScanState.scanStates) {
-    //     ku_dynamic_cast<RelDataReadState*>(
-    //         scanState.bwdScanState->dataScanState.get())
-    //         ->randomAccess = true;
-    //     auto relTablePair = relTables.find(tableID);
-    //     if (relTablePair != relTables.end()) {
-    //         scan(nodeID, relTablePair->second, onDiskScanState, *scanState.bwdScanState, result);
-    //     }
-    // }
-    return result;
-}
-
-bool OnDiskGraphScanState::InnerIterator::next() {
-    return tableScanState->source != TableScanSource::NONE &&
-           relTable->scan(context->getTx(), *tableScanState) && dstSelVector().getSelSize() > 0;
+bool OnDiskGraphScanState::InnerIterator::next(evaluator::ExpressionEvaluator* predicate) {
+    // TODO(Guodong/Ben): do we need to set selVector to unfiltered before scanning?
+    // I don't think it's necessary because scan doesn't read SelVector.
+    if (tableScanState->source == TableScanSource::NONE ||
+        !relTable->scan(context->getTx(), *tableScanState)) {
+        return false;
+    }
+    if (predicate != nullptr) {
+        predicate->select(tableScanState->outState->getSelVectorUnsafe());
+        tableScanState->outState->getSelVectorUnsafe().setToFiltered();
+    }
+    return dstSelVector().getSelSize() > 0;
```

Comment:
This function should only return false once the scan is finished, as it indicates to `Graph::Iterator` that the iterator is exhausted. You should probably put this in a loop and keep scanning until the result after the predicate is nonempty. Always returning true on this line should also work, but would be a bit slower as it will just process the empty data and call next again.

Reply:
Good catch. This is done.